Add background color selection in ``warp_perspective``

rvorias opened the pull request:
#### Changes
Added `"fill"` padding to warp_perspective().
This requires a tensor of shape 3 to be given as well, cf. arg `"padding_tensor"`
Shape of 3 is fixed for jit to work and if I'm not mistaken its prototypical use case is for filling
the background with an RGB color. (It could be extended to single channel and RGBA, but also
ran in some trouble with jit there)

Fixes #901 

#### Type of change
- [x] 📚  Documentation Update
- [x] 🔬 New feature (non-breaking change which adds functionality)

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code # also tested in notebook environment
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings # 4 tests failing for `test/enhance/test_histogram.py `, not related?
- [x] Did you update CHANGELOG in case of a major change? # n/a

edgarriba commented on kornia/geometry/transform/imgwarp.py:
```diff
@@ -56,7 +57,9 @@ def warp_perspective(
         dsize: size of the output image (height, width).
         mode: interpolation mode to calculate output values ``'bilinear'`` | ``'nearest'``.
-        padding_mode: padding mode for outside grid values ``'zeros'`` | ``'border'`` | ``'reflection'``.
+        padding_mode (str): padding mode for outside grid values
```
```suggestion
        padding_mode: padding mode for outside grid values
```

rvorias replied:
Removed `(str)`

edgarriba commented on kornia/geometry/transform/imgwarp.py:
```diff
@@ -36,6 +36,7 @@ def warp_perspective(
     dsize: Tuple[int, int],
     mode: str = 'bilinear',
     padding_mode: str = 'zeros',
+    padding_tensor: torch.Tensor = torch.zeros(3),  # needed for jit
```
```suggestion
    padding_tensor: Optional[torch.Tensor] = torch.zeros(3),  # needed for jit
```
because in below you check if it's `None`

rvorias replied:
This will give 
```
E           RuntimeError: 
E           'Optional[Tensor]' object has no attribute or method 'shape'.:
E             File "/Users/raphaelvorias/projects/kornia/kornia/geometry/transform/imgwarp.py", line 99
E               # fill padding is only supported for 3 channels because we can't set padding_tensor default
E               # to None as this gives jit issues.
E               if padding_mode == "fill" and padding_tensor.shape != torch.Size([3]):
E                                             ~~~~~~~~~~~~~~~~~~~~ <--- HERE
E                   raise ValueError(f"Padding_tensor only supported for 3 channels. Got {padding_tensor.shape}")
```

shijianjian commented on kornia/geometry/transform/imgwarp.py:
```diff
@@ -56,7 +57,9 @@ def warp_perspective(
-        padding_mode: padding mode for outside grid values ``'zeros'`` | ``'border'`` | ``'reflection'``.
+        padding_mode (str): padding mode for outside grid values
+          ``'zeros'`` | ``'border'`` | ``'reflection'`` | ``'fill'``.
+        padding_tensor: tensor of shape :math:`(3)` that fills the padding area. Only supported for RGB.
```
probably fill_value or something instead? padding_tensor reads confusing to me.

rvorias replied:
Updated to `fill_value`

edgarriba commented on kornia/geometry/transform/imgwarp.py:
```diff
@@ -102,7 +113,15 @@ def warp_perspective(
+        inv_ones_mask = 1 - F.grid_sample(ones_mask, grid, align_corners=align_corners, mode=mode, padding_mode="zeros")
+        inv_color_mask = inv_ones_mask * padding_tensor.unsqueeze(-1).unsqueeze(-1)
+        return F.grid_sample(src, grid, align_corners=align_corners, mode=mode, padding_mode="zeros") + inv_color_mask
+    else:
```
```suggestion
```
and un-identate below

rvorias replied:
Refactored

edgarriba commented on kornia/geometry/transform/imgwarp.py:
```diff
@@ -102,7 +113,15 @@ def warp_perspective(
+    if padding_mode == "fill":
+        # warp a mask of ones, then multiple with padding_tensor and add to src warp
+        ones_mask = torch.ones_like(src)
+        padding_tensor = padding_tensor.to(ones_mask.device)
```
```suggestion
        padding_tensor = padding_tensor.to(ones_mask)
```
both `device` and `dtype`

rvorias replied:
Updated

edgarriba reviewed (changes requested):
we should add some numerical tests to verify results here
https://github.com/kornia/kornia/blob/master/test/geometry/transform/test_imgwarp.py#L175

as side note: possibly this extra step should be abstracted as `_warp_and_fill` so that can be used by `warp_affine` too /cc @ducha-aiki

rvorias commented:
Thanks for the review remarks, will work on it this evening.

rvorias commented:
I've refactored the fill feature into `_warp_and_fill` as suggested. Now have to write some numerical tests.

rvorias commented:
Removed None assert and implemented two numerical tests for warp_perspective and warp_affine. All tests passed (but still waiting for test/x/test_image_classification.py to pass)

rvorias reviewed (commented):
self-review done

edgarriba commented on kornia/geometry/transform/imgwarp.py:
```diff
@@ -173,9 +189,38 @@ def warp_affine(
+    ones_mask = torch.ones_like(src)
+    fill_value = fill_value.to(ones_mask)[None, :, None, None]  # cast and add dimensions for broadcasting
+    inv_ones_mask = 1 - F.grid_sample(ones_mask, grid, align_corners=align_corners, mode=mode, padding_mode="zeros")
+    print(inv_ones_mask.shape, fill_value.shape)
```
```suggestion
```

edgarriba commented on kornia/geometry/transform/imgwarp.py:
```diff
@@ -173,9 +189,38 @@ def warp_affine(
     return F.grid_sample(src, grid, align_corners=align_corners, mode=mode, padding_mode=padding_mode)
 
 
+def _warp_and_fill(
```
```suggestion
def _fill_and_warp(
```
thinking it well - actually better name since that's the order of execution

dribnet commented:
Excited this has been merged - is this upstream from a planned release?
